### ISSUE TITLE
Dynamic App Title

### ISSUE THREAD

**Issue by gigiobix** [appreciated/vaadin-app-layout#241]
Hi,
First of all this is a wonderful layout!!!

It would be appreciated if it would be possible to set dynamically the App title based on viewed page.

Thanks in advance
Luigi

**Comment by Springrbua** [comment]
This should already be possible.
The LeftLayouts have a `setTitleComponent` so you only need to get access to the `AppLayout`. As much as I know you should be able to get it from the VaadinSession under the `AppLayout`-Token.

**Comment by appreciated** [comment]
@gigiobix Can you give an example how the API is supposed to support this?

**Comment by gigiobix** [comment]
I think it would be a good Idea to use something like the Page title mechanism via HasTitle Interface.
I.e. if the component which use your layout as Router  inherit from some Interface exposing getTitle method then the layout could invoke this method to set the title evrerytime the page is rendered.

Another way could be the one exposed by @Springrbua, get a reference to the Applayout and use the setTitleComponent method 

Another good Idea would be the ability to change the title during user interaction

I hope this could help your work.
Thanks

**Comment by Springrbua** [comment]
It would probably be a combination of both methods.
As I've said, the `AppLayout` already has a [`setTitleComponent`](https://github.com/appreciated/vaadin-app-layout/blob/master/app-layout-addon/src/main/java/com/github/appreciated/app/layout/behaviour/AppLayoutElementBase.java#L22), so changing the title should be pretty easy.
But the `AppLayoutRouterLayout` could implement an `Interface` (`AppLayoutContent`?), which has `default` methods to set the title etc.:
```
default void setAppLayoutTitleComponent(Component title) {
    UI.getCurrent().getSession().getAttribute(AppLayout.class).setTitleComponent(title);
}
```

**Comment by gigiobix** [comment]
Thanks to post of @Springrbua I've implemented a workaround to have dynamic title.
In the class extending `AppLayoutRouterLayout`, which I use as layout of all my routes, first to invoke the `init` method a set the `Applayout` generated in the session:

```
AppLayout layout = AppLayoutBuilder.get(Behaviour.LEFT_OVERLAY)
                                   .withTitle("GenericName")
                                   .build();

UI.getCurrent().getSession().setAttribute(AppLayout.class, layout);

init(layout);
```

Then in every class that use this layout after constructor, the I arrange the view I get the class from the `Session` and set the title
```
Optional.ofNullable(UI.getCurrent()
                      .getSession()
                      .getAttribute(AppLayout.class)).ifPresent(l -> {
			logger.info("Aggiorno il titolo a 'Gestione Processo'");
			 Span span = new Span("New Title");
		        span.setWidth("100%");
		        span.getStyle()
		                .set("white-space", "nowrap")
		                .set("overflow", "hidden")
		                .set("text-overflow", "ellipsis");
		        l.setTitleComponent(span);
		});
```
I hope this can help.

**Comment by Springrbua** [comment]
@gigiobix I have just seen, that the `AppLayoutRouterLayoutBase` (base-class for `AppLayoutRouterLayout`) has a static method [`getCurrent`](https://github.com/appreciated/vaadin-app-layout/blob/master/app-layout-addon/src/main/java/com/github/appreciated/app/layout/router/AppLayoutRouterLayoutBase.java#L31), which gives you the current `AppLayout` instance from the session.
So you could simply use `getCurrent()` if you are in a subclass of `AppLayoutRouterLayout`, or `AppLayoutRouterLayout.getCurrent()` from anywhere else.

**Comment by gigiobix** [comment]
@Springrbua , I've tried your suggestion, and it works fine.
I think it would be cool if `AppLayoutRouterLayoutBase` could expose the metod `void setTitle (String title)` for semplicity
Thanks a lot

**Comment by cedricaebi** [comment]
@Springrbua Sorry I'm still a huge noob. Can you provide a sample code for this? 
When I'm calling `AppLayoutRouterLayout.getCurrent()` at the end of my ViewConstructer, which uses the AppLayout from here as a parent i get a nullpointer.

**Comment by gigiobix** [comment]
@cedricaebi, I've tried thr method suggested by @Springrbua , but the method always return null value. 
To solve the problem, I use this tecnique:
At the end of layout creation, I store the result of build method in session:
```
AppLayout layout = 
		(AppLayoutBuilder.get(Behaviour.LEFT)
		 .withTitle("My Title")
		 .withAppMenu(menuBuilder.build())
		 .build());
UI.getCurrent().getSession().setAttribute(AppLayout.class, layout);
```
Then I created an Helper class with this static method, called to set view title:
```
	public static void setTitle(String title) {
		Span span = new Span(title);
		span.setWidth("100%");
		span.getStyle() 
			.set("white-space", "nowrap")
			.set("overflow", "hidden")
			.set("text-overflow", "ellipsis");
		Optional.ofNullable(UI.getCurrent().getSession().getAttribute(AppLayout.class)).ifPresent(l -> l.setTitleComponent(span));
	}
}
```
I hope this could help you.

**Comment by cedricaebi** [comment]
@gigiobix thanks! It worked. I called `Helper.setTitle()` in the `public void beforeEnter()` of the view.
Again thanks :)

**Comment by Springrbua** [comment]
@cedricaebi @gigiobix does this work if you have multiple tabs?
As much as I know, multiple tabs share the same Session and therefore the same attributes. But the AppLayout instance should not be the same as they have different DOMs.

**Comment by gigiobix** [comment]
@Springrbua our app is for internal usage, only one tab per time is opened, so I don't care about multi tab. I've tried to use the solution you suggested to me but it didn't function. If you can suggest a new solution I will try it.
Thanks

**Comment by Springrbua** [comment]
@gigiobix @cedricaebi 
I just tested the solution I provided and it works as expected.
Here the code:
```
public class MainView extends AppLayoutRouterLayout {
    public MainView() {
        init(AppLayoutBuilder.get(Behaviour.LEFT_RESPONSIVE)
            .withTitle("Test")
            .withAppBar(AppBarBuilder.get().add(new Button(new Icon(VaadinIcon.USER))).build())
            .withAppMenu(LeftAppMenuBuilder.get()
                .add(new LeftClickableItem("Test", VaadinIcon.ABACUS.create(), click -> {
                    getCurrent().setTitleComponent(new Span("Clicked"));
                }))
                .build()
            ).build()
        );
    }
    @Route(value = "", layout=MainView.class)
    public static class InnerView extends Div {
        protected void onAttach(AttachEvent attachEvent) {
            setText("Content");
        }
    }
}
```

The important line is inside the `LeftCLickableItem`s `ClickListener`:
`getCurrent().setTitleComponent(new Span("Clicked"));`
Outside the `AppLayoutRouterLayout` you hsould be able to do the same by calling `AppLayoutRouterLayout.getCurrent().setTitleComponent(...)`.

I also tested it with multiple tabs and noticed that it works **only for the last tab**.
The reason is probably the following:
Each browser tab shares the same `VaadinSession` and therefore the same `Attribute`s.
The last tab you opened writes it's `AppLayout` into the attribute and is therefore the only tab that can modify it.
Since all other tabs don't have this `AppLayout` instance attached, they are not allowed to change it.

So I guess we should really look for another solution and remove the `AppLayoutRouterLayout.getCurrent`.
The only solution I can think about right now is to go through all children of the `UI`, until we find the `AppLayoutRouterLayout` and return it.
@appreciated what do you think about this?

**Comment by gigiobix** [comment]
@Springrbua I'm not enough expert on Vaadin, but your thought on UI seems to be the right one, Perhaps it would be possible to assign an unique ID to the Title Container and refer to that instead of navigating the DOM.
I how this could help you

**Comment by appreciated** [comment]
> @appreciated what do you think about this?

This is a bug, and a misthinking of mine. Currently working on a fix for this.

**Comment by appreciated** [comment]
> 
> 
> @Springrbua , I've tried your suggestion, and it works fine.
> I think it would be cool if `AppLayoutRouterLayoutBase` could expose the metod `void setTitle (String title)` for semplicity
> Thanks a lot

To avoid a bloating of the  `AppLayoutRouterLayoutBase` with methods that do not belong in there, I disagree in adding a `setTitle`. Actually even `getCurrent` does not belong here. I have to think about this how to refactor this properly. Also since this is a breaking change this will be not be added in 2.* but in 3.0.0

**Comment by appreciated** [comment]
>  Actually even getCurrent does not belong here. I have to think about this how to refactor this properly. Also since this is a breaking change this will be not be added in 2.* but in 3.0.0
Opened #264.

> Each browser tab shares the same VaadinSession and therefore the same Attributes.
The last tab you opened writes it's AppLayout into the attribute and is therefore the only tab that can modify it.
Since all other tabs don't have this AppLayout instance attached, they are not allowed to change it.
Closed by ffd3372205c3df4e3f7c5aa74c48b1f82911da4a.
